### PR TITLE
Velocity Controllers

### PR DESCRIPTION
**Pull request by davetcoleman**
This PR is a little messy because it has this https://github.com/ros-controls/ros_controllers/pull/62 mixed in - once we merge that one this will look better. I'm just submitting this now so that I don't loose track of it/others know its mostly ready.

I've implemented a velocity hardware interface adapter for the `joint_trajectory_controller` that uses a simple PID loop for each joint to track input position and velocity and output a velocity command.

I've also implemented a velocity controller that takes as input position. Its basically a copy of the `effort_controller/joint_position_controller`.

I've tested both of these on Baxter.

### CODE REVIEW

File: joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h [hunk 1]
```diff
@@ -104,13 +131,112 @@ class HardwareInterfaceAdapter<hardware_interface::PositionJointInterface, State
...
+    // Preconditions
+    if (!joint_handles_ptr_)
+      return;
+    assert(n_joints == state_error.position.size());
+    assert(n_joints == state_error.velocity.size());
+
+    // Update PIDs
+    for (unsigned int i = 0; i < n_joints; ++i)
+    {
+      const double command = pids_[i]->computeCommand(state_error.position[i], state_error.velocity[i], period);
```
Comment (adolfo-rt) [inline joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h]: Have you tried adding the `desired_state.velocity` as a feed-forward component to the command?. In the current implementation the PID sees both the reference command and the tracking error as perturbations to compensate for, which will probably require larger gains. If OTOH you add a feed-forward `desired_state.velocity` (just add the term to the above line), the perturbations the PID has to compensate become much smaller, hence will require smaller gains.

Comment (davetcoleman) [inline joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h]: Excellent point, that improved my joint controllers R^2 position tracking error by an average of 2.8%!

Comment (davetcoleman) [inline joint_trajectory_controller/include/joint_trajectory_controller/hardware_interface_adapter.h]: It is now

```
const double command = pids_[i]->computeCommand(state_error.position[i], desired_state.velocity[i], period);
```

**Comment by fmessmer** [comment]
I tried to split the various features in separate pull requests:
#108 velocity_controller/JointPositionController
#109 velocity interface for JointTrajectoryController
#110 joint_mode_controller

**Comment by davetcoleman** [comment]
Sorry for not being more active on this, but I am not currently using ros_control and am working on entirely different (and virtual) projects.

**Comment by davetcoleman** [comment]
I'll assume these split pull requests can replace this PR. Thanks for all your effort @ipa-fxm

**Pull request closed without merging**